File: dynamic-programming/CoinChange.java
import java.util.Arrays;

/**
 * Problem:
 * You are given an integer array coins representing coins of different denominations and an integer amount
 * representing a total amount of money. Return the fewest number of coins that you need to make up that amount.
 * If that amount of money cannot be made up by any combination of the coins, return -1.
 * You may assume that you have an infinite number of each kind of coin.
 */
public class CoinChange {

    // Approach 1: Memoization (Top-Down)
    // Time Complexity: O(amount * n), where n is the number of coins.
    // Space Complexity: O(amount) for the memoization table + O(amount) for recursion stack.
    public int coinChangeMemo(int[] coins, int amount) {
        int[] memo = new int[amount + 1];
        int result = solve(coins, amount, memo);
        return (result == Integer.MAX_VALUE) ? -1 : result;
    }

    private int solve(int[] coins, int target, int[] memo) {
        if (target == 0) {
            return 0;
        }
        if (target < 0) {
            return Integer.MAX_VALUE;
        }
        if (memo[target] != 0) {
            return memo[target];
        }

        int minCoins = Integer.MAX_VALUE;
        for (int coin : coins) {
            int res = solve(coins, target - coin, memo);
            if (res != Integer.MAX_VALUE) {
                minCoins = Math.min(minCoins, 1 + res);
            }
        }
        memo[target] = minCoins;
        return minCoins;
    }

    // Approach 2: Tabulation (Bottom-Up)
    // Time Complexity: O(amount * n)
    // Space Complexity: O(amount) for the DP table.
    public int coinChangeTab(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1); // Use a value larger than amount
        dp[0] = 0;

        for (int i = 1; i <= amount; i++) {
            for (int coin : coins) {
                if (i - coin >= 0) {
                    dp[i] = Math.min(dp[i], 1 + dp[i - coin]);
                }
            }
        }
        return (dp[amount] > amount) ? -1 : dp[amount];
    }

    public static void main(String[] args) {
        CoinChange coinChanger = new CoinChange();
        int[] coins = {1, 2, 5};
        int amount = 11;

        System.out.println("Fewest coins for amount " + amount + " (Memoization): " + coinChanger.coinChangeMemo(coins, amount));
        System.out.println("Fewest coins for amount " + amount + " (Tabulation): " + coinChanger.coinChangeTab(coins, amount));

        int[] coins2 = {2};
        int amount2 = 3;

        System.out.println("Fewest coins for amount " + amount2 + " (Memoization): " + coinChanger.coinChangeMemo(coins2, amount2));
        System.out.println("Fewest coins for amount " + amount2 + " (Tabulation): " + coinChanger.coinChangeTab(coins2, amount2));

        int[] coins3 = {1};
        int amount3 = 0;

        System.out.println("Fewest coins for amount " + amount3 + " (Memoization): " + coinChanger.coinChangeMemo(coins3, amount3));
        System.out.println("Fewest coins for amount " + amount3 + " (Tabulation): " + coinChanger.coinChangeTab(coins3, amount3));
    }
}

File: dynamic-programming/coin_change.py
# Coin Change - Dynamic Programming

"""
Problem:
You are given an integer array coins representing coins of different denominations and an integer amount
representing a total amount of money. Return the fewest number of coins that you need to make up that amount.
If that amount of money cannot be made up by any combination of the coins, return -1.
You may assume that you have an infinite number of each kind of coin.
"""

import sys

class CoinChange:
    # Approach 1: Memoization (Top-Down)
    # Time Complexity: O(amount * n), where n is the number of coins.
    # Space Complexity: O(amount) for the memoization table + O(amount) for recursion stack.
    def coinChangeMemo(self, coins: list[int], amount: int) -> int:
        memo = {}

        def solve(target):
            if target in memo:
                return memo[target]
            if target == 0:
                return 0
            if target < 0:
                return sys.maxsize

            min_coins = sys.maxsize
            for coin in coins:
                res = solve(target - coin)
                if res != sys.maxsize:
                    min_coins = min(min_coins, 1 + res)
            
            memo[target] = min_coins
            return min_coins

        result = solve(amount)
        return result if result != sys.maxsize else -1

    # Approach 2: Tabulation (Bottom-Up)
    # Time Complexity: O(amount * n)
    # Space Complexity: O(amount) for the DP table.
    def coinChangeTab(self, coins: list[int], amount: int) -> int:
        dp = [sys.maxsize] * (amount + 1)
        dp[0] = 0

        for i in range(1, amount + 1):
            for coin in coins:
                if i - coin >= 0 and dp[i - coin] != sys.maxsize:
                    dp[i] = min(dp[i], 1 + dp[i - coin])
        
        return dp[amount] if dp[amount] != sys.maxsize else -1

# Test cases
if __name__ == "__main__":
    coin_changer = CoinChange()
    coins = [1, 2, 5]
    amount = 11
    
    print(f"Fewest coins for amount {amount} (Memoization): {coin_changer.coinChangeMemo(coins, amount)}")
    print(f"Fewest coins for amount {amount} (Tabulation): {coin_changer.coinChangeTab(coins, amount)}")

    coins = [2]
    amount = 3
    
    print(f"Fewest coins for amount {amount} (Memoization): {coin_changer.coinChangeMemo(coins, amount)}")
    print(f"Fewest coins for amount {amount} (Tabulation): {coin_changer.coinChangeTab(coins, amount)}")

    coins = [1]
    amount = 0
    
    print(f"Fewest coins for amount {amount} (Memoization): {coin_changer.coinChangeMemo(coins, amount)}")
    print(f"Fewest coins for amount {amount} (Tabulation): {coin_changer.coinChangeTab(coins, amount)}")
